### DISCUSSION TITLE
how to open the GCLog ？

### DISCUSSION THREAD

**Discussion by shushanhf** [dotnet/runtime#93858] [Q&A]
Hi, @Maoni0 
How to open the GCLog ?
For example the runtime6.0,  
I modify the `src/coreclr/gc/gcconfig.h`
```
+    BOOL_CONFIG  (LogEnabled,             "GCLogEnabled",           NULL,                             true,             "Specifies if you want to turn on logging in GC")  
```

Also open the macros within the `gc/gcpriv.h`
```
+#define TRACE_GC
+#define SIMPLE_DPRINTF
```

```
export DOTNET_GCLogEnabled=1
export DOTNET_GCLogFileSize=100
export DOTNET_GCLogFile="/home/qiao/GCLogQiao.log"
export DOTNET_StressLog=0
#export DOTNET_gcConcurrent=0
#export DOTNET_GCName="libclrgc.so"
export DOTNET_GCRegionsRange=20000000
export DOTNET_heapverify=1
```

But the created file  `/home/qiao/GCLogQiao.log.log`  is empty.

**Comment by Maoni0** [comment]
you'll need to enable logging in the code first - uncomment these 2 lines in gcpriv.h -

```cplusplus
#define TRACE_GC
#define SIMPLE_DPRINTF
```
this is not compiled in by default because this is the internal GC log and is mostly only meant to be used for active GC development.

**Reply by shushanhf** [reply]
Sorry, I did't get it.
After uncomment the two lines
```
#define TRACE_GC
#define SIMPLE_DPRINTF
```

Which env value should be opened ? The `export DOTNET_GCLogEnabled=1`  or the `DOTNET_LogEnable=1` and `DOTNET_LogToConsole=1` ?
But the last two are not the GC's info.

**Comment by Maoni0** [comment]
you had them already -

export DOTNET_GCLogEnabled=1
export DOTNET_GCLogFileSize=100
export DOTNET_GCLogFile="/home/qiao/GCLogQiao.log"
export DOTNET_StressLog=0

**Reply by shushanhf** [reply]
Thanks.
But there is no `/home/qiao/GCLogQiao.log.log` created after these values opened.

Liking these lines
https://github.com/dotnet/runtime/blob/4a336b1c4756fe95274658c15bc9a4674625a8b2/src/coreclr/gc/gc.cpp#L12493-L12499

The macro ` #ifdef TRACE_GC`  is not defined. The GCLogFile will not be created.

**Comment by Maoni0** [comment]
you mean TRACE_GC is not defined? 
it should be defined in gcpriv.h if you uncomment those 2 lines I mentioned above.

**Reply by shushanhf** [reply]
But if defined the `TRACE_GC`, the file `/home/qiao/GCLogQiao.log.log` is created only as an empty log file.

**Reply by Maoni0** [reply]
you need to define TRACE_GC or the code to create a log wouldn't be there. this is what I shared with another customer who needed to get the dprintf logs: [trace_gc](https://github.com/dotnet/runtime/compare/main...Maoni0:runtime:trace_gc). this enables logging code and shows how to change the level so you can log just the selected few things.

have you done any GCs? if not, the log is still in memory. we always flush the in-memory buffer to the file in do_post_gc (see `flush_gc_log (false);`). make sure you are actually doing GCs. if you want to see what's in the in-memory buffer, you can dump what's at `WKS::gc_log_buffer` (current offset is `WKS::gc_log_buffer_offset`). replace WKS with SVR if you are using Server GC. see the (simple) implementation in `log_va_msg`. I would suggest to do a little debugging if this doesn't work.
